### ISSUE TITLE
Decide on code embed system for deep dives

### ISSUE THREAD

**Issue by schalkneethling** [mdn/yari#4139]
As part of a deep-dive article, we will want to show some code examples. During the research phase, we used Codepen for these.

Are we going to continue with this approach or will we instead use plain `code` blocks with syntax highlighting handled by Prism?

If we do decide to continue using Codepen, should we get a security review on their usage from the security team? I know in the past we did need this when we wanted to use codepen for the Web API interactive examples.

**Comment by Rumyra** [comment]
I think this actually ties in with how we present demos in the future on regular mdn. Myself and @wbamberg have it in mind we'll start discussing this this quarter. I did a bit of a hack a couple of weekends ago to see if you could make a codepen style editor reading from existing html/css/js files: https://github.com/remotehack/remotepen (actually pretty straight forward).

The reason for this was we can keep the demo code as a standalone demo (for regular mdn a lot of the demos are files in other repositories, but this way they could easily sit in content or plus).

- API
- big example
  - Interface
    - Method
      - small example ...

It was through this hack that someone then pointed me in the direction of https://indiepen.tech/ which looks like a pretty viable solution, maybe a customisation.

Although I am a fan of codepen, and believe we could talk to them to figure out a good embed workflow, we do have issues surrounding contributers updating codepens (they would more than likely need access) - and while the two requirements for regular vs plus might differ, I personal like the idea of one solution here 👍

**Comment by Rumyra** [comment]
Also @schalkneethling it _might_ be easier to style both the embed _and_ the demo itself this way 🤷🏻‍♀️

**Comment by schalkneethling** [comment]
Also note that the reason I brought this up now is that, during the last round, the security team did have some concerns over Codepen's observatory score. Another concern they raised was tracking. It is not entirely clear what is pulled into a page when you embed a Codepen and so, they wanted clarity on the kinds of tracking that is done.

**Comment by Rumyra** [comment]
Yeh the more I think about it the more I like it - wonder if we could chat to the original makers (it's totally open source _but_) I would assume we'd want to change the style and functionality a bit.

The other thing to I consider worth a note (which is where some of the trickiness lies with one solution fits all) is not _all_ demos have one html, one css and one js file (think web workers, modules, css only demos) - however I do think starting with a solution which covers most bases is probably a good approach ☺️

**Comment by schalkneethling** [comment]
> Yeh the more I think about it the more I like it - wonder if we could chat to the original makers (it's totally open source _but_) I would assume we'd want to change the style and functionality a bit.

Indeed, perhaps we can help with some upstream work to make it themeable. Perhaps they already have an API that one can use to extend some of the functionality that just is not documented yet? Would be great to reach out and find out. I would love to support a project like this, and I feel that it fits very well into Mozilla's wheelhouse to do just that.

> The other thing to I consider worth a note (which is where some of the trickiness lies with one solution fits all) is not _all_ demos have one html, one css and one js file (think web workers, modules, css only demos) - however I do think starting with a solution which covers most bases is probably a good approach ☺️

Agreed. Also, for those more complicated things, perhaps we can look at something like Glitch?

**Comment by peterbe** [comment]
Some drive-by comments;

1. Nothing beats a simple `<pre>` block (presumably created by a triple-backtick Markdown file) in terms of reliability and performance. 
2. Let's not assume that we'll have Kumascript rendering in Deep-dives. There are very serious implications of introducing a private GitHub repo into our build pipeline. So let's not assume we can rely on `{{EmbedLiveSample}}` in deep-dives. Or `{{EmbedInteractiveExample(...)}}`.
3. This Indiepen looks nice but is it reliable? It might work today but A) how uptime is it? B) what if they decide to let the project die due to lack of interest/funds? 
4. What @schalkneethling said about the importance of asking for permission first from Mozilla Sec people.

**Comment by Rumyra** [comment]
Other thoughts:

> Nothing beats a simple <pre> block (presumably created by a triple-backtick Markdown file) in terms of reliability and performance.

True, but we have multiple languages for demos and codepen-esq things work well for this

> Let's not assume that we'll have Kumascript rendering in Deep-dives. There are very serious implications of introducing a private GitHub repo into our build pipeline. So let's not assume we can rely on {{EmbedLiveSample}} in deep-dives. Or {{EmbedInteractiveExample(...)}}

I was considering the editor as an alternative - we still need to find a good way to tidy up webAPI demos and the macros have never been suitable for these, so there's a mash of codepens and fiddles and github repos

> This Indiepen looks nice but is it reliable? It might work today but A) how uptime is it? B) what if they decide to let the project die due to lack of interest/funds?

In my mind I was thinking of approaching so we could 'take' the code and make it our own, rather than forking. (We keep a 'powered by indiepen' link or something) I agree we should have ownership over the solution (for the same reasons)

**Comment by schalkneethling** [comment]
> In my mind I was thinking of approaching so we could 'take' the code and make it our own, rather than forking. (We keep a 'powered by indiepen' link or something) I agree we should have ownership over the solution (for the same reasons)

Does Indiepen not depend on Github as the backend? i.e. you point it at a repo that uses a specific layout and from that it generates pens?

**Comment by Rumyra** [comment]
To sum up what was just discussed:

- We should look into two options:
  - Roll our own (iterate on the interactive examples we have currently) 
  - Pick up the discussion to use Codepen

- We should probably create a pros & cons for each based on the requirements we have:
  - Offline capabilities
  - Security
  - Ease of updating code for the examples
  - (Personal) I would like something that also worked for WebAPI demos (HTML, CSS & JS) as we have littered codepens/jsfiddles etc... and it'd be good to take the opportunity to find a solution for those too

Please add anything else to the list above you can think of.

Mentioning @dalexsy so she can see this 👍

**Comment by peterbe** [comment]
Performance! It matters a lot for SEO. If Google thinks our pages are slow. we get less traffic (aka. eye balls). And people don't like pages that feel sluggish and slow too.

One of the reasons the current interactive examples are so slow is because they ship with [codemirror](https://codemirror.net/) which is a full editor and not just a syntax highlighter. I don't know what Codepen does but it probably has a similar solution. 

If Codepen is better, go for it but it'd have to be carefully tested that they don't make the pages slower.

**Comment by peterbe** [comment]
By the way, I think an iframe is still a sane thing to do but one thing Gregor and I often speculated about was to NOT make it "interactive" unless you have to. Instead of loading the very heavy `codemirror` we just syntax highlight it and *if* the person wants to interactive with it by doing edits to see what happens, only then do we load codemirror.

**Comment by Rumyra** [comment]
Update:

We have three loose categories for code and examples displayed on MDN:

1. Static code (think `<pre>` blocks)
2. Interactive examples (HTML, CSS, JS)
3. Other more broad demo code, which includes multiple files in multiple languages (web api examples, dom examples) - most of these are currently in various repositories in the mdn org

For the redesign we keep 1 and 2 as is, just looking good. 3. needs more consideration and could possibly affect plus because deep dives could contain demos which have multiple files to see the final result.

Ideally we want an interactive in page code editor (like codepen) for these examples, however there are a lot of considerations for this. It could become its own standalone project outside of this scope.

Some notes:
- If we don't look at a codepen style editor until later, how do we manage examples authors create that include multiple files (deep dive related)
- Should we take the opportunity to move 2. Interactive examples into content
- Should an interactive editor be available for 1. Static code at the request of a user (e.g an 'edit this code' button)
- We actually have a version of the interactive examples which incorporates HTML, CSS & JS, however it doesn't really work because of ShadowDOM issues

**Comment by peterbe** [comment]
> We have three loose categories for code and examples displayed on MDN:

I would say 4. 

1. `<pre>` blocks 
2. interactive examples (that uses mdn/bob and mdn/interactive-examples) 
3. GitHub (mdn.github.io) iframes
4. Live samples (part of the content, gets rendered in an iframe)

**Comment by peterbe** [comment]
By the way, an example of a GitHub embed live sample is https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Position_skills which uses for its iframe https://mdn.github.io/css-examples/learn/tasks/position/position2.html

**Comment by Rumyra** [comment]
Maybe there's 5 - because we don't want to forget all the demos in the gh repos which are not live samples or pulled out into iframes (for whatever reason)

**Comment by wbamberg** [comment]
There are also examples that need a server-side component, which currently don't work at all on MDN since we retired the samples-server (https://github.com/mdn/content/discussions/5772).

And it's worth distinguishing between live samples that are really just examples, and ones that include their own editor UI and are thus more like interactive examples (e.g. in the Learn area, but all the way up to complete little web apps like https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Background_and_Borders/Border-image_generator ).
